grin/src/AbstractInterpretation/LiveVariable/CodeGen.hs

```diff
@@ -196,53 +239,55 @@ codeGenM e = (cata folder >=> const setMainLive) e
-      -- NOTE: A function might have side-effects,
-      -- so we have to generate code for it even if its result register is dead.
-      -- emit $ funResultReg `isLiveThen` bodyInstructions
+        R r -> varPatternDataFlow funResultReg r
       pure Z
 
     EBindF leftExp lpat rightExp -> do
-      leftExp >>= \case
-        Z -> case lpat of
-          Unit -> pure ()
-          Var name -> do
-            r <- newReg
-            addReg name r
-          _ -> error $ "pattern mismatch at LVA bind codegen, expected Unit got " ++ show lpat
-        R r -> case lpat of
-          Unit  -> setBasicValLive r
-          Lit{} -> setBasicValLive r
-          Var name -> do
-            varReg <- newReg
-            addReg name varReg
-            varPatternDataFlow varReg r
-          ConstTagNode tag args -> do
-            irTag <- getTag tag
-            setTagLive irTag r
-            bindInstructions <- codeGenBlock_ $ forM (zip [1..] args) $ \(idx, arg) ->
-              case arg of
-                Var name -> do
-                  argReg <- newReg
-                  addReg name argReg
-                  nodePatternDataFlow argReg r irTag idx
-                Lit {} -> emit IR.Set { dstReg = r, constant = IR.CNodeItem irTag idx live }
-                _ -> error $ "illegal node pattern component " ++ show arg
-            emit IR.If
-              { condition     = IR.NodeTypeExists irTag
-              , srcReg        = r
-              , instructions  = bindInstructions
-              }
-          _ -> error $ "unsupported lpat " ++ show lpat
-      rightExp
+      lhs <- leftExp
+      let R lhsReg = lhs
+
+      case lpat of
+      {- NOTE: By convention, all bindings should have a variable pattern
+          or a simple left-hand side of form `pure <var>`. This guarantees
+          that all relevant computations will have a name. Also, it means
+          the information has been already propagated to the variable.
+      -}
+        Unit  -> setBasicValLive lhsReg
+        Lit{} -> setBasicValLive lhsReg
+        Var v -> do
+          varReg <- newReg
+          addReg v varReg
+          varPatternDataFlow varReg lhsReg
+        ConstTagNode tag args -> do
+          irTag <- getTag tag
+          setTagLive irTag lhsReg
+          bindInstructions <- codeGenBlock_ $ forM (zip [1..] args) $ \(idx, arg) ->
+            case arg of
+              Var name -> do
+                argReg <- newReg
+                addReg name argReg
+                nodePatternDataFlow argReg lhsReg irTag idx
+              Lit {} -> emit IR.Set { dstReg = lhsReg, constant = IR.CNodeItem irTag idx live }
+              _ -> error $ "illegal node pattern component " ++ show arg
+          emit IR.If
+            { condition     = IR.NodeTypeExists irTag
+            , srcReg        = lhsReg
+            , instructions  = bindInstructions
+            }
+        _ -> error $ "unsupported lpat " ++ show lpat
+
+      rhs <- rightExp
```

Comment:
What about `rhs@(R rhsRef) <- rightRxp` and `pure rhs` at the end?

Reply:
The problem is that there is no `MonadFail` instance for the code gen monad.

Reply:
Ah got it. I am not sure which is the better, to define a MonadFail for pattern match errors, or throw a runtime exception on pattern match error :) Leave it as it is.